docs/testing/unit-tests.md

```diff
@@ -1,3 +1,160 @@
+
+
+## Folder structure
+Keep your tests in a `test` folder in your working directory. The test file should have the same name as the test subject file, and use the extension `.js`.
+
+```
++-- test
+|   +-- index.js
++-- index.jsx
+```
+
+### Describing tests
+Use 'describe' block to group test cases. Each test case should ideally describe one behaviour only.
+
+In test cases, try to describe in plain words the expected behaviour For UI components, this might entail describing expected behaviour from a user perspective rather than explaining code internals.
+
+**Good**
+
+```javascript
+describe( 'CheckboxWithLabel', () => {
+    test('checking checkbox should disable the form submit button', () => {
+        ...
+    } );
+} );
+```
+
+**Not so good**
+
+```javascript
+describe( 'CheckboxWithLabel', () => {
+    test('checking checkbox should set this.state.disableButton to `true`', () => {
+        ...
+    } );
+} );
+```
+
+### Snapshot testing
+See[Snapshot testing](snapshot-testing.md)
+
+### Examples
+
+#### Mocking HTTP requests
+
+[Nock](https://github.com/node-nock/nock) allows us to override HTTP requests, and unit test code in isolation. For example, you might have an action  that returns a GET request in a thunk:
+
+```javascript
+// action.js
+export function requestSomeData() {
+	return dispatch => {
+		return wpcom
+            .getSomeData()
+            .then( data => {
+				dispatch( {
+					type: SOME_DATA_SUCCESS,
+					data,
+				} );
+			} )
+			.catch( error => {
+				dispatch( {
+					type: SOME_DATA_FAILURE,
+					error,
+				} );
+			} );
+	};
+}
+```
+
+To mock this response, we can specify the nock parameters in a describe block and use `persist()` to intercept all requests in that block.
+
+Here's a simple example:
+
+```javascript
+import { requestSomeData } from '../actions';
+import useNock from 'test/helpers/use-nock';
+
+describe( 'requestSomeData', () => {
+	describe( 'successful response', () => {
+		useNock( nock => {
+			nock( 'https://public-api.wordpress.com:443' )
+				.persist()
+				.get( '/rest/v1.1/some-endpoint' )
+				.reply( 200, { hoo: 'ray!' } );
+		} );
+
+		test( 'dispatches success action object when successful request is completed', () => {
+			const mockDispatchCallback = jest.fn();
+            // Trigger the action to return the thunk
+			requestSomeData()( mockDispatchCallback ).then( () => {
+                expect( mockDispatchCallback.mock.calls[ 0 ][ 0 ] ).toEqual( {
+                    type: SOME_DATA_SUCCESS,
+                    data: { hoo: 'ray!' }
+                } );
+            } );
+		} );
+     } );
+} );
+```
+
+### Mocking dependencies
```

Comment:
This is a good opportunity to help out with some common problems (e.g. `store`, `lib/analytics`) but `lib/wp` should get a special mention. We've got several different clever approaches worth distilling.

Reply:
Yeah it's hard to know how far to take it. It's easy find an example of how to, say, mock the config object:

```
jest.mock( 'config', () => ( {
	isEnabled: jest.fn(),
} ) );
```

But then, I guess it would be helpful to provide a working example, since the code snippets do vary.

Reply:
Personally, I would discourage mocking whenever possible 😄

Reply:
I'm fine with mentioning `store`, `lib/analytics` or `lib/wp`, but in most of the cases they have corresponding code that is side-effects free:

* `analytics` - can be rewritten with Redux actions - https://github.com/Automattic/wp-calypso/blob/master/client/state/analytics/actions.js#L93
*  `wp` - can be rewritten using data layer - https://github.com/Automattic/wp-calypso/tree/master/client/state/data-layer

We should rather put it in its own section called `Legacy code` or something along those lines.